### PR TITLE
Follow XDG config dir spec in new mc installations

### PR DESCRIPTION
**Pull request by vadmeste**
config.json & certs directories will be in ~/.config/mc/ and share
downloads.json and uploads.json will be in ~/.local/share/mc/share/

This is draft because I am not sure if we want this.

### CODE REVIEW

**Comment by ddevault** [comment]
>This is draft because I am not sure if we want this.

What's the cause for hesitation?

File: cmd/config.go [hunk 1]
```diff
@@ -48,13 +48,24 @@ func getMcConfigDir() (string, *probe.Error) {
 	if e != nil {
 		return "", probe.NewError(e)
 	}
-	configDir := filepath.Join(homeDir, defaultMCConfigDir())
-	return configDir, nil
+	if runtime.GOOS == "windows" {
+		return filepath.Join(homeDir, mcDirName()), nil
+	}
+	// Check if ~/.mc already exists, deprecated Sep 2021
+	oldMcConfigDir := filepath.Join(homeDir, mcDirName())
```
Comment (klauspost) [inline cmd/config.go:55]: `mcDirName` no longer returns the same.

File: cmd/config.go [hunk 2]
```diff
@@ -48,13 +48,24 @@ func getMcConfigDir() (string, *probe.Error) {
...
+	if st, err := os.Stat(oldMcConfigDir); err == nil && st.IsDir() {
+		return oldMcConfigDir, nil
+	}
+	// Construct mc config dir path according to https://specifications.freedesktop.org/basedir-spec/basedir-spec-0.6.html
+	xdgPath := os.Getenv("XDG_CONFIG_HOME")
+	if xdgPath == "" {
+		xdgPath = filepath.Join(homeDir, ".config")
```
Comment (klauspost) [inline cmd/config.go:62]: Just return the previous path `~/.mc`. No need to have even more places to look.

File: cmd/share.go [hunk 3]
```diff
@@ -105,13 +107,30 @@ func shareSetColor() {
...
+	// Check if ~/.mc already exists, deprecated Sep 2021
+	oldMcConfigDir := filepath.Join(homeDir, "."+mcDirName())
+	if st, err := os.Stat(oldMcConfigDir); err == nil && st.IsDir() {
+		return filepath.Join(oldMcConfigDir, globalSharedURLsDataDir), nil
+	}
+
+	// Construct mc config dir path according to https://specifications.freedesktop.org/basedir-spec/basedir-spec-0.6.html
+	xdgPath := os.Getenv("XDG_DATA_HOME")
+	if xdgPath == "" {
+		xdgPath = filepath.Join(homeDir, ".local/share")
```
Comment (klauspost) [inline cmd/share.go:131]: Use previous location. No need for a third.

File: cmd/config.go (hunk 2 above)
Comment (vadmeste) [reply cmd/config.go:62]: > Just return the previous path `~/.mc`. No need to have even more places to look.

I am following the specification here, if XDG_CONFIG_HOME is not defined, then you need to use ~/.config/mc/

Comment (ddevault) [reply cmd/config.go:62]: Using ~/.config/mc is the correct behavior in this case. The only time ~/.mc should be used is if the config file cannot be found at a better location but exists in ~/.mc, i.e. for installs which pre-date this change.

Comment (klauspost) [reply cmd/config.go:62]: I don't really care about the specification, I just don't want the config to potentially be in 3 different places. 2 is confusing enough.

Comment (ddevault) [reply cmd/config.go:62]: No one else really cares about the spec, either, but to do the right thing you need to set the fallback to ~/.config/mc for users who don't have the environment variable set. Most users do not have the environment variable set and the problem will not be resolved to their satisfaction without this logic.

**Comment by harshavardhana** [comment]
> config.json & certs directories will be in ~/.config/mc/ and share
> downloads.json and uploads.json will be in ~/.local/share/mc/share/

@vadmeste let's not do this as the default unless XDG is available and set - this is not true for macOS, BSD systems. 

XDG is a wider standard than just this you have to follow the following basics to keep properly support for XDG

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

- There is a single base directory relative to which user-specific data files should be written. This directory is defined by the environment variable $XDG_DATA_HOME.

- There is a single base directory relative to which user-specific configuration files should be written. This directory is defined by the environment variable $XDG_CONFIG_HOME.

- There is a single base directory relative to which user-specific state data should be written. This directory is defined by the environment variable $XDG_STATE_HOME.

- There is a single base directory relative to which user-specific executable files may be written.

- There is a set of preference ordered base directories relative to which data files should be searched. This set of directories is defined by the environment variable $XDG_DATA_DIRS.

- There is a set of preference ordered base directories relative to which configuration files should be searched. This set of directories is defined by the environment variable $XDG_CONFIG_DIRS.

- There is a single base directory relative to which user-specific non-essential (cached) data should be written. This directory is defined by the environment variable $XDG_CACHE_HOME.

- There is a single base directory relative to which user-specific runtime files and other file objects should be placed. This directory is defined by the environment variable $XDG_RUNTIME_DIR.

Our code needs to honor these basics, look for each of these ENVs and apply them appropriately. 

Also NOTE: using `.local/mc` and `.config/mc` will conflict with `midnight-commander` so you need to be careful about choosing the same config path.

**Comment by vadmeste** [comment]
> Also NOTE: using `.local/mc` and `.config/mc` will conflict with `midnight-commander` so you need to be careful about choosing the same config path.

Yes, that's why I am not too keen in supporting XDG here. This looks like only important if we want to add mc to GNU/Linux distributions.

**Comment by ddevault** [comment]
The binary also conflicts with midnight-commander, which predates minio. An executable name collision is much worse than a config directory collision, why should it apply to this change but not the other?

**Comment by harshavardhana** [comment]
> The binary also conflicts with midnight-commander, which predates minio. An executable name collision is much worse than a config directory collision, why should it apply to this change but not the other?

https://github.com/minio/mc/blob/master/CONFLICT.md

**Comment by harshavardhana** [comment]
> > Also NOTE: using `.local/mc` and `.config/mc` will conflict with `midnight-commander` so you need to be careful about choosing the same config path.
> 
> Yes, that's why I am not too keen in supporting XDG here. This looks like only important if we want to add mc to GNU/Linux distributions.

Let's not do this for now then - this is not an important requirement for now. We can add it only if it solves some real needs.

**Comment by ddevault** [comment]
What nonsense.

**Comment by klauspost** [comment]
Closed as per #3627

**Pull request closed without merging**